src/replication.c

```diff
@@ -2308,6 +2474,441 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+    sdsfree(err);
+    goto error;
+}
+
+/* Replication: Replica side.
+ * Initialize server.pending_repl_data infrastructure, we will allocate the buffer
+ * itself once we need it */
+void replDataBufInit(void) {
+    serverAssert(server.pending_repl_data.blocks == NULL);
+    server.pending_repl_data.len = 0;
+    server.pending_repl_data.peak = 0;
+    server.pending_repl_data.blocks = listCreate();
+    server.pending_repl_data.blocks->free = zfree;
+}
+
+/* Replication: Replica side.
+ * Track the local repl-data buffer streaming progress and serve clients from time to time */
+void replStreamProgressCallback(size_t offset, int readlen, time_t *last_progress_callback) {
+    time_t now = mstime();
+    if (server.loading_process_events_interval_bytes &&
+        (offset + readlen) / server.loading_process_events_interval_bytes >
+            offset / server.loading_process_events_interval_bytes &&
+        now - *last_progress_callback > server.loading_process_events_interval_milliseconds) {
+        replicationSendNewlineToPrimary();
+        processEventsWhileBlocked();
+        *last_progress_callback = now;
+    }
+}
+
+/* Replication: Replica side.
+ * Reads replication data from primary into specified repl buffer block */
+int readIntoReplDataBlock(connection *conn, replDataBufBlock *o, size_t read) {
+    int nread = connRead(conn, o->buf + o->used, read);
+    if (nread == -1) {
+        if (connGetState(conn) != CONN_STATE_CONNECTED) {
+            serverLog(LL_VERBOSE, "Error reading from primary: %s", connGetLastError(conn));
+            cancelReplicationHandshake(1);
+        }
+        return C_ERR;
+    }
+    if (nread == 0) {
+        if (server.verbosity <= LL_VERBOSE) {
+            serverLog(LL_VERBOSE, "Provisional primary closed connection");
+        }
+        cancelReplicationHandshake(1);
+        return C_ERR;
+    }
+    o->used += nread;
+    server.stat_total_reads_processed += nread;
+    return read - nread;
+}
+
+/* Replication: Replica side.
+ * Returns true in case the replica's local repl-baffer used all of its space */
+int isReplicaBufferLimitReached(void) {
+    return server.pending_repl_data.len > server.client_obuf_limits[CLIENT_REPLICA].hard_limit_bytes;
+}
+
+/* Replication: Replica side.
+ * Read handler for buffering incoming repl data during RDB download/loading. */
+void bufferReplData(connection *conn) {
+    size_t readlen = PROTO_IOBUF_LEN;
+    int read = 0;
+
+    while (readlen > 0) {
+        listNode *ln = listLast(server.pending_repl_data.blocks);
+        replDataBufBlock *tail = ln ? listNodeValue(ln) : NULL;
+
+        /* Append to tail string when possible */
+        if (tail && tail->used < tail->size) {
+            size_t avail = tail->size - tail->used;
+            read = min(readlen, avail);
+            readlen -= read;
+            read = readIntoReplDataBlock(conn, tail, read);
```

Comment:
It is a bit unconventional for a "read" function to return the remaining buf length. If we would like to keep the contract, which is fine with me, consider renaming `read` to `remaining_size` or something similar
```suggestion
            remaining_size = readIntoReplDataBlock(conn, tail, read);
```

Reply:
👍 renaming to `remaining_bytes`